beluga_amcl/src/amcl_node.cpp

```diff
@@ -816,19 +818,67 @@ void AmclNode::timer_callback() {
     return;
   }
 
+  // Particle weights from the filter may or may not be representative of the
+  // true distribution. If we resampled, they are not, and there will be multiple copies
+  // of the most likely candidates, all with unit weight. In this case the number of copies
+  // is a proxy for the prob density at each candidate. If we did not resample before updating
+  // the estimation and publishing this message (which can happen if the resample interval
+  // is set to something other than 1), then all particles are expected to be different
+  // and their weights are proportional to the prob density at each candidate.
+  //
+  // Only the combination of both the state distribution and the candidate weights together
+  // provide information about the probability density at each candidate.
+  // To handle both cases, we group repeated candidates and compute the accumulated weight
+
+  struct RepresentativeData {
+    Sophus::SE2d state;
+    double weight{0.};
+  };
+
+  struct RepresentativeBinHash {
+    std::size_t operator()(const Sophus::SE2d& s) const noexcept {
+      std::size_t h1 = std::hash<double>{}(s.translation().x());
+      std::size_t h2 = std::hash<double>{}(s.translation().y());
+      std::size_t h3 = std::hash<double>{}(s.so2().log());
+      return h1 ^ (h2 << 1) ^ (h3 << 2);
+    }
+  };
+
+  struct RepresentativeBinEqual {
+    bool operator()(const Sophus::SE2d& lhs, const Sophus::SE2d& rhs) const noexcept {
+      // good enough, since copies of the same candidate are expected to be identical copies
+      return lhs.translation().x() == rhs.translation().x() &&  //
+             lhs.translation().y() == rhs.translation().y() &&  //
+             lhs.so2().log() == rhs.so2().log();
+    }
+  };
+
+  std::unordered_map<Sophus::SE2d, RepresentativeData, RepresentativeBinHash, RepresentativeBinEqual>
+      representatives_map;
+  representatives_map.reserve(particle_filter_->particle_count());
+  double max_weight = 1e-5;  // never risk dividing by zero
+
+  for (const auto& [state, weight] :
+       ranges::views::zip(particle_filter_->states_view(), particle_filter_->weights_view())) {
+    auto& representative = representatives_map[state];  // if the element does not exist, create it
+    representative.state = state;
+    representative.weight += weight;
+    if (representative.weight > max_weight) {
+      max_weight = representative.weight;
+    }
+  }
```

Comment:
@glpuga meta: the ability to retrieve a normalized probability distribution from the particle filter sounds useful. Should this functionality be part of the filter itself? Or at least a core utility.

Reply:
I thought about it, but from what I saw, unless the estimation algorithm reconstructs the pdf to make an estimation, there little use for this for the particle filter itself. It's also kind of an expensive luxury, https://en.wikipedia.org/wiki/Kernel_density_estimation .

RViz2, on the other hand, would probably benefit from this as a better and more clear representation that those currently available (arrow, other arrows, and triads).